src/service/resource.js

```diff
@@ -287,29 +306,32 @@ export class Resource {
   /**
    * Requests the resource's element to be built. See {@link AmpElement.build}
    * for details.
+   * @return {?Promise}
    */
   build() {
-    if (this.blacklisted_ || !this.element.isUpgraded()
-        || !this.resources_.grantBuildPermission()) {
-      return;
-    }
-    try {
-      this.element.build();
-    } catch (e) {
-      dev().error(TAG, 'failed to build:', this.debugid, e);
+    if (this.isBuilding_ ||
+        this.blacklisted_ ||
+        !this.element.isUpgraded() ||
+        !this.resources_.grantBuildPermission()) {
+      return null;
```

Comment:
Right now it is - as part of `Resources` class. I'm currently refactoring it to eliminate this, but it will take me some time to address build part. I did add some TODOs. If you'd like me to add some additional checks - I will happily do. I'm not too far off otherwise: I've almost removed `blacklisted_` and pretty close to removing `grantBuildPermission` - I'll be able to remove  this whole clause then. But I wanted to go with this in the meantime since it blocks the downstream work.